### PR TITLE
Add a loss comparison script

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #2049
* __->__ #2029

## Summary
This PR adds `scripts/loss_compare.py` for comparing training losses between different git commits and/or training configurations.

## Key Features

  - Commit Comparison: Compare losses between two different git commits with deterministic training
  - Configuration Comparison: Compare different training configurations on the same commit
  - Reproducibility: Automatically enables deterministic mode and seed checkpointing for reproducible
  comparisons
  - Real-time Output: Streams training output to both console and log files during execution
  - Statistical Analysis: Generates step-by-step loss comparisons and summary statistics
  - CI Testing: Includes --assert-equal flag for automated testing to verify identical losses

#### Compare two commits
```
python3 ./scripts/loss_compare.py main my_branch
```
#### Compare two commits with custom configuration 
```
python3 ./scripts/loss_compare.py main my_branch \
--baseline-config="./custom.toml" 
--baseline-options="--parallelism.tensor_parallel_degree=2"  \
```

#### Compare different parallelization strategies on same commit
```
python3 ./scripts/loss_compare.py . . \
--baseline-config="./llama3_8b.toml" 
--baseline-options="--parallelism.tensor_parallel_degree=2" \
--test-options="--parallelism.tensor_parallel_degree=1" \
```

#### Assert equality for CI testing
```
python3 ./scripts/loss_compare.py main my_branch --assert-equal
```

## Real Use Cases
Compare full dtensor simple fsdp with fsdp2:
```
python3 scripts/loss_compare.py . . \
--baseline-options='--activation_checkpoint.mode="none"'  \
--test-train-file='torchtitan.experiments.full_dtensor.train' \ 
--test-options='--model.name full_dtensor.llama3 --activation_checkpoint.mode="none"'  \
 --assert-equal --no-seed-checkpoint


[LOSS_COMPARE]
[LOSS_COMPARE] Asserting losses are equal...
[LOSS_COMPARE] Baseline log: /tmp/baseline_training.log
[LOSS_COMPARE] Test log: /tmp/test_training.log
[LOSS_COMPARE] Extracted 100 steps from baseline log
[LOSS_COMPARE] Extracted 100 steps from test log
test_losses_equal (__main__.assert_losses_equal.<locals>.LossEqualityTest.test_losses_equal) ... ok
```

### CODE REVIEW

File: .github/workflows/integration_test_8gpu_features.yaml [hunk 1]
```diff
@@ -76,5 +76,11 @@ jobs:
         export TEST_WITH_ROCM=$([[ "${{ matrix.gpu-arch-type }}" == "rocm" ]] && echo 1 || echo 0)
         python -m tests.integration_tests.run_tests --test_suite features $RUNNER_TEMP/artifacts-to-be-uploaded --ngpu 8
 
+        # Verify the accuracy.
+        export baseline_options="--parallelism.data_parallel_replicate_degree=1"
+        export test_options="--parallelism.data_parallel_replicate_degree=4"
```
Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu_features.yaml]: Why comparing these two? FSDP 8 vs. HSDP 4, 2

Also we are not using the assert_equal flag, so what happens if the losses are different? Are you analyzing them with some other scripts?

I thought the idea was we should compare main and PR submit, but for that it may be easier to assert with a fixed loss number.

Comment (fegin) [reply .github/workflows/integration_test_8gpu_features.yaml]: Yes, I should add `--assert_equal`—I'll update it.

I originally planned to compare the current PR with the main branch. The problem is that CI mounts the folder as read-only, so I couldn't check out the "main" version. I took a step back and decided to verify the minimum guarantees: at least some parallelism variations should have the same losses. Since this check is quick, I think it doesn't hurt to add it (we can also verify TP). This way, we ensure the minimum guarantees and also validate the script.

For comparing with the "main" branch, one way to achieve this is to pre-record the results and add an option to the tool to read them. Both baseline and test runs must match the pre-recorded losses when `--assert_equal` is set. The downside is that this approach is less stable than simply checking out during CI, since the pre-recorded method can't tolerate machine or library changes.

I prefer to do this in another PR.

Comment (tianyu-l) [reply .github/workflows/integration_test_8gpu_features.yaml]: > since the pre-recorded method can't tolerate machine or library changes.

Good point, but here are two cases if numerics change?
1. not "error", just differences between machines / libraries
2. library introduces a bug which causes both main and branch on all parallelisms to be the same and simultaneously incorrect.

I guess we could say the latter should be captured by upstream library?

File: .github/workflows/integration_test_8gpu_features.yaml [hunk 2]
```diff
@@ -77,9 +77,10 @@ jobs:
         python -m tests.integration_tests.run_tests --test_suite features $RUNNER_TEMP/artifacts-to-be-uploaded --ngpu 8
 
         # Verify the accuracy.
+        echo "Checking FSDP4 v.s. HSDP2FSDP2TP2 accuracy parity"
         export baseline_options="--parallelism.data_parallel_replicate_degree=1"
-        export test_options="--parallelism.data_parallel_replicate_degree=4"
-        python3 scripts/loss_compare.py . . --baseline-options="${baseline_options}" --test-options="${test_options}" --steps=10 --job-dump-folder="${RUNNER_TEMP}/artifacts-to-be-uploaded/accuracy_comparison_outputs"
+        export test_options="--parallelism.data_parallel_replicate_degree=2 --parallelism.tensor_parallel_degree=2"
+        python3 scripts/loss_compare.py . . --baseline-options="${baseline_options}" --test-options="${test_options}" --steps=10 --job-dump-folder="${RUNNER_TEMP}/artifacts-to-be-uploaded/accuracy_comparison_outputs" --assert-equal --baseline-ngpus=4 --test-ngpus=8 --steps=1
```
Comment (tianyu-l) [inline .github/workflows/integration_test_8gpu_features.yaml]: Would this be able to pass the failure to CI? In the past we've seen errors not propagated to CI, so we weren't be able to detect failures.

Comment (fegin) [reply .github/workflows/integration_test_8gpu_features.yaml]: Didn't that happened before. My guess is yes as I saw the error earlier. But let's test it out.

Comment (fegin) [reply .github/workflows/integration_test_8gpu_features.yaml]: Yes, the error propagates correctly. https://github.com/pytorch/torchtitan/actions/runs/19486677666/job/55770375464?pr=2029

File: .github/workflows/integration_test_8gpu_features.yaml (hunk 1 above)
Comment (fegin) [reply .github/workflows/integration_test_8gpu_features.yaml]: Sure, we can do this. Will do this in another PR.

**Review by wconstab — APPROVED** [review]
Looks useful! One thing I noticed is you run git checkout without checking the git status is clean first. It can fail to checkout if there are files in the working directory that git would overwrite during the checkout. It can also be annoying if git forces you to clean your directory and you have logs or things in there, but I personally think it's worth it

**Comment by fegin** [comment]
@wconstab Good point, I'll add the git status check as well.
